### PR TITLE
add read.opensearch(reconstruct_document =True) option

### PR DESCRIPTION
**Pull request by baitsguy**
Adds a reconstruct_document flag to the opensearch reader. This allows you to query an opensearch index and receive a DocSet in it's pre-exploded form. This is useful to do original document level transforms on a DocSet.

Added a "_doc_source" property to be able to identify which elements were loaded as query hits, and which were loaded as part of document reconstruction.

Eventually we might want to move this as a separate transform but this gives us some immediate value and can be moved out later.

### CODE REVIEW

**Review by HenryL27 — APPROVED** [review]
agree on moving it to its own transform at some point... this seems pretty tightly coupled to the OpenSearch Reader though so wondering if you have thoughts about how to split it. 
Also I guess we're gonna want to keep the implode flag in the read interface though I guess that's as simple as sticking a `if implode, return docset.implode()` or something.

approved with some suggestions.

ps thanks for the OpenSearchReaderQueryResponse unittests

File: lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py [hunk 1]
```diff
@@ -71,18 +73,102 @@ def check_target_presence(self, query_params: BaseDBReader.QueryParams):
...
+        all_elements = []
+        for i in range(0, len(doc_ids), batch_size):
+            doc_ids_batch = doc_ids[i : i + batch_size]
+            from_offset = 0
+            while True:
+                query = {
+                    "query": {"terms": {"parent_id.keyword": doc_ids_batch}},
+                    "size": page_size,
+                    "from": from_offset,
+                }
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py:163]: Yeah I figured we can use it if the code gets complex, but better to keep pagination state out of opensearch if we can

File: lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py [hunk 2]
```diff
@@ -71,18 +73,102 @@ def check_target_presence(self, query_params: BaseDBReader.QueryParams):
...
+                        **element.get("_source", {}),
+                    }
+                )
+                assert doc.parent_id, "Got non-element record from OpenSearch reconstruction query"
+                if doc.doc_id not in query_result_elements_per_doc.get(doc.parent_id, {}):
+                    doc.properties[DocumentPropertyTypes.SOURCE] = DocumentSource.DOCUMENT_RECONSTRUCTION_RETRIEVAL
+                else:
+                    doc.properties[DocumentPropertyTypes.SOURCE] = DocumentSource.DB_QUERY
+                parent = unique_docs[doc.parent_id]
+                parent.elements.append(Element(doc.data))
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py:140]: Will do as a follow up to https://github.com/aryn-ai/sycamore/pull/864, we need element_ids

File: lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py [hunk 3]
```diff
@@ -71,18 +73,102 @@ def check_target_presence(self, query_params: BaseDBReader.QueryParams):
...
+                doc = Document(
+                    {
+                        **element.get("_source", {}),
+                    }
+                )
+                assert doc.parent_id, "Got non-element record from OpenSearch reconstruction query"
+                if doc.doc_id not in query_result_elements_per_doc.get(doc.parent_id, {}):
+                    doc.properties[DocumentPropertyTypes.SOURCE] = DocumentSource.DOCUMENT_RECONSTRUCTION_RETRIEVAL
+                else:
+                    doc.properties[DocumentPropertyTypes.SOURCE] = DocumentSource.DB_QUERY
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply lib/sycamore/sycamore/connectors/opensearch/opensearch_reader.py:138]: https://github.com/aryn-ai/sycamore/pull/864 uses the former

File: lib/sycamore/sycamore/data/document.py [hunk 4]
```diff
@@ -1,12 +1,24 @@
...
 
+class DocumentSource(Enum):
+    UNKNOWN: str = "UNKNOWN"
+    DB_QUERY: str = "DB_QUERY"
+    DOCUMENT_RECONSTRUCTION_RETRIEVAL: str = "DOCUMENT_RECONSTRUCTION_RETRIEVAL"
+
+
+class DocumentPropertyTypes:
+    SOURCE: str = "_doc_source"
+    PAGE_NUMBER: str = "page_number"
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply lib/sycamore/sycamore/data/document.py:19]: Will update partitioner as a follow up to https://github.com/aryn-ai/sycamore/pull/864, will do it there

**Pull request merged**